Update Gateway API to v0.7.0

meyskens opened the pull request:
This PR updates our Gateway API to v0.7.0. This includes the new vendor changes as well as all fixes we needed to pass the additional conformance tests in individual commits.


Fixes: https://github.com/cilium/cilium/issues/25474

```release-note
Support Gateway API v0.7.0
```

meyskens committed:
Vendor in Gateway API v0.7.0

This change updates our Gateway API to use v0.7.0
It also has a fix where the type was changed from a string in upstream.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Implement programmed flag

This status condition is now required to be set for Gateway API
conformance.
This status is now set once a gateway has valid config and has an
assidned address

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Add AttachedRoutes

This adds a count for the attachedroutes field in the gateway status.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Set Accepted status on listeners

Setting gatewayListenerAcceptedCondition is now required to pass
conformance

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Add logic to remove listener status

The new conformance tests now explicitly test for removed listeners to
be removed from the status.
We did not do this before, this adds a filter for that.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Set refernece resolved to true when ok

This change sets the resolved reference status to true when they have
been found.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Check for section match on gateway

when a section is demanded on a gateway reference we should check if it
actually exists. The conformace tests now check against this.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Rename cert reference allowed function

This renames the function to check reference grants for certificates to
a less generic name to avoid confusion with the one responsible for
checking route backends.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Check ReferenceGrants to backends in status validation

We previously did not perform checks for reference grants when
validating routes. This now extracts this check into a helper which is
re-used in both ingestion code as well as in the route validation code.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Break in validation to not accept partial invalid routes

Previously we would check all references. This would cause the status of
the reference to be overritten in the last one was correct.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Generate manifests with new kubebuilder

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Enable HTTPRoute in conformance

HTTPRoute is no logner a standard enabled feature in the conformace
tests, This adds this back in our CI tests

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

sayboras commented on operator/pkg/gateway-api/httproute.go:
```diff
@@ -92,6 +93,8 @@ func (r *httpRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		// Watch for changes to Backend services
 		Watches(&source.Kind{Type: &corev1.Service{}}, r.enqueueRequestForBackendService()).
+		// Watch for changes to Reference Grants
+		Watches(&source.Kind{Type: &gatewayv1beta1.ReferenceGrant{}}, r.enqueueRequestForRequestGrant()).
```
:100:

sayboras reviewed (approved):
Thanks for working on this, also cleaning up my bad code/mistake before :+1:

meyskens committed:
Implement ReferenceGrants correct for TLSRoute

This PR adds correct ReferenceGrant checks into TLSRoute. It also adds a
check for resource kind in the helper function.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

viktor-kurchenko commented on operator/pkg/gateway-api/gateway_status.go:
```diff
@@ -45,6 +51,29 @@ func gatewayStatusAcceptedCondition(gw *gatewayv1beta1.Gateway, accepted bool, m
 }
 
+func gatewayStatusProgrammedCondition(gw *gatewayv1beta1.Gateway, scheduled bool, msg string) metav1.Condition {
+	switch scheduled {
```
Just wondering, why `switch` not `if`?

meyskens replied:
This is an adaptation from @sayboras's original code, not sure what his reasoning was :)

sayboras replied:
ah i might have started with some enum, and then refactored to bool, so switch might be just lazy refactor.

meyskens replied:
I have had an idea last night to re-factor this into 1 generic blob to re-use. But that is probably out of scope for this PR.

viktor-kurchenko commented on operator/pkg/gateway-api/helpers/referencegrants.go:
```diff
@@ -0,0 +1,38 @@
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+// isReferenceAllowed returns true if the reference is allowed by the reference grant.
```
`isReferenceAllowed` -> `IsBackendReferenceAllowed`

meyskens replied:
Will change this in the next PR where i also add checks for certificates

meyskens committed:
Improve func argument order

Improve the function arguments to the checked route is the last one.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Add IsSecretReferenceAllowed

This adds IsSecretReferenceAllowed to check the secret reference logic
in one place.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Improve error messages on ReferenceGrant fetch

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

youngnick commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -143,14 +144,16 @@ func (r *gatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return fail(err)
 	}
 
+	setGatewayReady(gw, true, "Gateway successfully configured")
```
Hmm, Ready is now deprecated, actually, so we shouldn't set it.

meyskens replied:
Removing it :)

youngnick commented on operator/pkg/gateway-api/gateway_reconcile.go:
```diff
@@ -319,8 +349,10 @@ func (r *gatewayReconciler) setListenerStatus(ctx context.Context, gw *gatewayv1
 			conds = append(conds, gatewayListenerInvalidRouteKinds(gw, "Invalid Route Kinds"))
+			isValid = false
 		} else {
 			conds = append(conds, gatewayListenerProgrammedCondition(gw, true, "Listener Ready"))
```
Ready has been deprecated as part of this release, so we should remove it.

youngnick reviewed (approved):
LGTM with one small thing about removing `Ready` if we can. It's only deprecated, so not a blocker, but if we can remove now, that would be better.

meyskens committed:
Remove deprecated ready status

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>
